devito/dse/aliases.py

```diff
@@ -36,244 +38,246 @@ def collect(exprs):
         ideal_anti_stencil = Stencil.union(*[i.anti_stencil for i in group])
         for i in group:
             if i.anti_stencil.subtract(ideal_anti_stencil).empty:
                 aliases[i.alias] = i.relax(ideal_anti_stencil)
 
-    return mapper, aliases
+    return aliases
 
 
 # Helpers
 
-def create_alias(expr, offsets):
-    """
-    Create an aliasing expression of ``expr`` by replacing the offsets of each
-    indexed object in ``expr`` with the new values in ``offsets``. ``offsets``
-    is an ordered sequence of tuples with as many elements as the number of
-    indexed objects in ``expr``.
-    """
-    indexeds = retrieve_indexed(expr, mode='all')
-    assert len(indexeds) == len(offsets)
+Candidate = namedtuple('Candidate', 'expr indexeds bases offsets')
 
-    mapper = {}
-    for indexed, ofs in zip(indexeds, offsets):
-        aindices = IterationInstance(indexed).aindices
-        assert len(aindices) == len(ofs)
-        mapper[indexed] = indexed.function[[i + j for i, j in zip(aindices, ofs)]]
 
-    return expr.xreplace(mapper)
+def analyze(expr):
+    """
+    Determine whether ``expr`` is a potential alias and collect relevant metadata.
 
+    A necessary condition is that all Indexeds in ``expr`` are affine in the
+    access Dimensions so that the access offsets (or "strides") can be derived.
+    For example, given the following Indexeds: ::
 
-def calculate_COM(offsets):
-    """
-    Determine the centre of mass (COM) in a collection of offsets.
+        A[i, j, k], B[i, j+2, k+3], C[i-1, j+4]
 
-    The COM is a basis to span the vectors in ``offsets``.
+    All of the access functions all affine in ``i, j, k``, and the offsets are: ::
 
-    Also return the distance of each element E in ``offsets`` from the COM (i.e.,
-    the coefficients that when multiplied by the COM give exactly E).
+        (0, 0, 0), (0, 2, 3), (-1, 4)
     """
-    COM = []
-    for ofs in zip(*offsets):
-        handle = []
-        for i in zip(*ofs):
-            strides = sorted(set(i))
-            # Heuristic:
-            # - middle point if odd number of values, or
-            # - strides average otherwise
-            index = int((len(strides) - 1) / 2)
-            if (len(strides) - 1) % 2 == 0:
-                handle.append(strides[index])
+    # No way if writing to a tensor or an increment
+    if expr.lhs.is_Indexed or expr.is_Increment:
+        return
+
+    indexeds = retrieve_indexed(expr.rhs, mode='all')
+    if not indexeds:
+        return
+
+    bases = []
+    offsets = []
+    for i in indexeds:
+        ii = IterationInstance(i)
+
+        # There must not be irregular accesses, otherwise we won't be able to
+        # calculate the offsets
+        if ii.is_irregular:
+            return
+
+        # Since `ii` is regular (and therefore affine), it is guaranteed that `ai`
+        # below won't be None
+        base = []
+        offset = []
+        for e, ai in zip(ii, ii.aindices):
```

Comment:
i know that would imply in a lot of changes all over the code (not only this PR)
but my internal voice kind of reads this line like:
`for expr, aindex in zip(iter, iter.aindices)`: